server/mailrepository/mailrepository-memory/src/main/java/org/apache/james/mailrepository/memory/MemoryMailRepository.java

```diff
@@ -0,0 +1,76 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailrepository.memory;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+
+import org.apache.james.mailrepository.api.MailRepository;
+import org.apache.mailet.Mail;
+
+public class MemoryMailRepository implements MailRepository {
+
+    private final HashMap<String, Mail> mails;
+
+    public MemoryMailRepository() {
+        mails = new HashMap<>();
```

Comment:
A concurrentHashMap look safer IMO.

Remember that mailetContainer is multi-threaded...

Reply:
that's funny because I tried to prove it's necessary before using ConcurrentHashMap but failed at that so didn't changed it.

Reply:
Yes, running concurent test on a single thread is generally not successful :trollface: